src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -308,6 +310,27 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
 
                 int lastRead;
 
+                // Read up to 3 bytes to see if it's the UTF-8 BOM
+                do
+                {
+                    // No need for checking for growth, the minimal rent sizes both guarantee it'll fit.
+                    Debug.Assert(rented.Length >= utf8Bom.Length);
+
+                    lastRead = stream.Read(
+                        rented,
+                        written,
+                        utf8Bom.Length - written);
+
+                    written += lastRead;
+                } while (lastRead > 0 && written < utf8Bom.Length);
+
+                // If we have 3 bytes, and they're the BOM, reset the write position to 0.
+                if (written == utf8Bom.Length &&
+                    utf8Bom.SequenceEqual(rented.AsSpan(0, utf8Bom.Length)))
+                {
+                    written = 0;
+                }
+
                 do
                 {
                     if (rented.Length == written)
```

Comment:
Just to confirm, this will never be true on the first iteration of the loop (if the stream contained a bom and we reset `written = 0`), correct?

Reply:
`rented` has a minimum size of 4 (and then, further, whatever the arraypool rounds up to from there).

If we don't reset, then it's (4+) vs 3, so not equal.  If we do reset then it's (4+) vs 0, so also not equal.